File: go-gin/internal/dto/requests.go
package dto

import (
	"time"

	"github.com/google/uuid"
	"github.com/rest-comparison/go-gin/internal/models"
)

type ErrorDetail struct {
	Field   string `json:"field"`
	Message string `json:"message"`
}

type ErrorResponse struct {
	Error   string        `json:"error"`
	Details []ErrorDetail `json:"details,omitempty"`
}

type ShipCreateRequest struct {
	Name        string           `json:"name" validate:"required"`
	Class       models.ShipClass `json:"class" validate:"required,oneof=cruiser frigate destroyer scout"`
	Registry    string           `json:"registry" validate:"required"`
	WarpCapable *bool            `json:"warp_capable"`
}

type ShipUpdateRequest struct {
	Name        *string           `json:"name" validate:"omitempty,min=1"`
	Class       *models.ShipClass `json:"class" validate:"omitempty,oneof=cruiser frigate destroyer scout"`
	Registry    *string           `json:"registry" validate:"omitempty,min=1"`
	WarpCapable *bool             `json:"warp_capable"`
}

type CrewmateCreateRequest struct {
	ShipID  uuid.UUID          `json:"ship_id" validate:"required"`
	Name    string             `json:"name" validate:"required"`
	Role    models.CrewmateRole `json:"role" validate:"required,oneof=captain engineer medic pilot gunner"`
	Species string             `json:"species" validate:"required"`
	Rank    int                `json:"rank" validate:"required"`
}

type CrewmateUpdateRequest struct {
	ShipID  *uuid.UUID          `json:"ship_id"`
	Name    *string             `json:"name" validate:"omitempty,min=1"`
	Role    *models.CrewmateRole `json:"role" validate:"omitempty,oneof=captain engineer medic pilot gunner"`
	Species *string             `json:"species" validate:"omitempty,min=1"`
	Rank    *int                `json:"rank"`
}

type MissionCreateRequest struct {
	ShipID    uuid.UUID            `json:"ship_id" validate:"required"`
	Codename  string               `json:"codename" validate:"required"`
	Objective string               `json:"objective" validate:"required"`
	Status    *models.MissionStatus `json:"status" validate:"omitempty,oneof=planned active completed failed"`
	StartedAt *time.Time           `json:"started_at"`
	EndedAt   *time.Time           `json:"ended_at"`
}

type MissionUpdateRequest struct {
	ShipID    *uuid.UUID            `json:"ship_id"`
	Codename  *string               `json:"codename" validate:"omitempty,min=1"`
	Objective *string               `json:"objective" validate:"omitempty,min=1"`
	Status    *models.MissionStatus `json:"status" validate:"omitempty,oneof=planned active completed failed"`
	StartedAt *time.Time            `json:"started_at"`
	EndedAt   *time.Time            `json:"ended_at"`
}

File: kotlin-springboot/src/main/kotlin/com/spaceship/dto/Requests.kt
package com.spaceship.dto

import com.fasterxml.jackson.annotation.JsonProperty
import com.spaceship.model.CrewmateRole
import com.spaceship.model.MissionStatus
import com.spaceship.model.ShipClass
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotNull
import java.time.Instant
import java.util.UUID

data class ShipCreateRequest(
    @field:NotBlank val name: String,
    @field:NotNull @field:JsonProperty("class") val shipClass: ShipClass,
    @field:NotBlank val registry: String,
    val warpCapable: Boolean? = null
)

data class ShipUpdateRequest(
    val name: String? = null,
    @field:JsonProperty("class") val shipClass: ShipClass? = null,
    val registry: String? = null,
    val warpCapable: Boolean? = null
) {
    fun isEmpty(): Boolean = name == null && shipClass == null && registry == null && warpCapable == null
}

data class CrewmateCreateRequest(
    @field:NotNull val shipId: UUID,
    @field:NotBlank val name: String,
    @field:NotNull val role: CrewmateRole,
    @field:NotBlank val species: String,
    @field:NotNull val rank: Int
)

data class CrewmateUpdateRequest(
    val shipId: UUID? = null,
    val name: String? = null,
    val role: CrewmateRole? = null,
    val species: String? = null,
    val rank: Int? = null
) {
    fun isEmpty(): Boolean = shipId == null && name == null && role == null && species == null && rank == null
}

data class MissionCreateRequest(
    @field:NotNull val shipId: UUID,
    @field:NotBlank val codename: String,
    @field:NotBlank val objective: String,
    val status: MissionStatus? = null,
    val startedAt: Instant? = null,
    val endedAt: Instant? = null
)

data class MissionUpdateRequest(
    val shipId: UUID? = null,
    val codename: String? = null,
    val objective: String? = null,
    val status: MissionStatus? = null,
    val startedAt: Instant? = null,
    val endedAt: Instant? = null
) {
    fun isEmpty(): Boolean =
        shipId == null && codename == null && objective == null && status == null && startedAt == null && endedAt == null
}
